### PR TITLE
boards: added support for nucleo-l476 (stm32l4)

### PR DESCRIPTION
**Pull request by haukepetersen**
I couldn't resist...

But somehow the UART is not working, I suspect the clock config code+ configuration for the board to still be broken, will take a deeper look tomorrow...

### CODE REVIEW

File: cpu/stm32l4/cpu.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    /* reset PLL config register */
+    RCC->PLLCFGR = 0;
+    /* set high speed clock as source for the PLL */
+    RCC->PLLCFGR |= CLOCK_PLL_SOURCE;
+    /* set division factor for main PLL input clock */
+    RCC->PLLCFGR |= (CLOCK_PLL_M & 0x3F);
+    /* set main PLL multiplication factor for VCO */
+    RCC->PLLCFGR |= (CLOCK_PLL_N & 0x1FF) << 6;
+    /* set main PLL division factor for main system clock */
+    RCC->PLLCFGR |= (((CLOCK_PLL_P & 0x03) >> 1) - 1) << 16;
```
Comment (aabadie) [inline cpu/stm32l4/cpu.c]: I think it should be R instead of P here ?

File: boards/nucleo-l476/include/board.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+extern "C" {
+#endif
+
+/**
+ * @name xtimer configuration
+ * @{
+ */
+#define XTIMER_DEV          TIMER_DEV(0)
+#define XTIMER_CHAN         (0)
+#define XTIMER_WIDTH        (16)
```
Comment (aabadie) [inline boards/nucleo-l476/include/board.h]: according to periph_conf, this should be 32

File: boards/nucleo-l476/include/periph_conf.h [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    {
+        .dev      = TIM5,
+        .max      = 0xffffffff,
+        .rcc_mask = RCC_APB1ENR1_TIM5EN,
+        .bus      = APB1,
+        .irqn     = TIM5_IRQn
+    }
+};
+
+#define TIMER_0_ISR         isr_tim2
```
Comment (aabadie) [inline boards/nucleo-l476/include/periph_conf.h]: looks like an issue

File: cpu/stm32l4/ldscripts/stm32l476rg.ld [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2017 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup      cpu_stm32l0
```
Comment (aabadie) [inline cpu/stm32l4/ldscripts/stm32l476rg.ld]: l4

File: cpu/stm32l4/ldscripts/stm32l476rg.ld [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * @{
+ *
+ * @file
+ * @brief           Memory definitions for the STM32L053R8
```
Comment (aabadie) [inline cpu/stm32l4/ldscripts/stm32l476rg.ld]: L476

**Comment by aabadie** [comment]
@haukepetersen, do you have the gpio ? (is LED0_ON working in main ?)

**Comment by haukepetersen** [comment]
@aabadie: Thanks for the review, will check now. And sorry for the bad state of the PR, had to leave yesterday in a hurry and wanted to publish this before I left...

**Comment by aabadie** [comment]
@haukepetersen, if you want, I have a 'cleanest' branch locally.

**Comment by aabadie** [comment]
but it doesn't work either

**Comment by haukepetersen** [comment]
I am just rewriting the clock init code (as the F4 code does not match at all), hope to have it working in some minutes..

**Comment by aabadie** [comment]
> hope to have it working in some minutes..

fine with me, can't wait for this board saying hello

File: cpu/stm32l4/include/periph_cpu.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#define PERIPH_CPU_H
+
+#include "periph_cpu_common.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Available ports on the STM32L0 family
```
Comment (aabadie) [inline cpu/stm32l4/include/periph_cpu.h]: L0

File: cpu/stm32l4/vectors.c [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+WEAK_DEFAULT void isr_lcd(void);
+WEAK_DEFAULT void isr_0(void);
+WEAK_DEFAULT void isr_rng(void);
+WEAK_DEFAULT void isr_fpu(void);
+
+/* interrupt vector table */
+ISR_VECTORS const void *interrupt_vector[] = {
+    /* Exception stack pointer */
+    (void*) (&_estack),             /* pointer to the top of the stack */
+    /* Cortex-M0 handlers */
```
Comment (aabadie) [inline cpu/stm32l4/vectors.c]: M4

**Comment by haukepetersen** [comment]
the clock init stuff should now be fine, though I still get no output via UART... Will further debug.

File: cpu/stm32l4/include/periph_cpu.h [hunk 8]
```diff
@@ -27,7 +27,7 @@ extern "C" {
 #endif
 
 /**
- * @brief   Available ports on the STM32L0 family
+ * @brief   Available port
```
Comment (aabadie) [inline cpu/stm32l4/include/periph_cpu.h]: port**s** ?

**Comment by haukepetersen** [comment]
got it, xtimer_msg, default, hello-world -> all runs fine now.

File: boards/nucleo-l476/include/board.h [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ * Copyright (C) 2017 Freie Universität Berlin
+ *               2017 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    boards_nucleo-l053 Nucleo-L053
```
Comment (aabadie) [inline boards/nucleo-l476/include/board.h]: l476

File: boards/nucleo-l476/include/board.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * @ingroup     boards
+ * @brief       Board specific files for the nucleo-l053 board
```
Comment (aabadie) [inline boards/nucleo-l476/include/board.h]: l476 (and just below as well)

File: boards/nucleo-l476/include/periph_conf.h [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+ * CORECLOCK        -> 80MHz MAX!
+ */
+#define CLOCK_PLL_M         (6)
+#define CLOCK_PLL_N         (20)
+#define CLOCK_PLL_R         (2)
+/* peripheral clock setup */
+#define CLOCK_AHB_DIV       RCC_CFGR_HPRE_DIV1
+#define CLOCK_AHB           (CLOCK_CORECLOCK / 1)
+#define CLOCK_APB1_DIV      RCC_CFGR_PPRE1_DIV4
+#define CLOCK_APB1          (CLOCK_CORECLOCK / 4)
```
Comment (aabadie) [inline boards/nucleo-l476/include/periph_conf.h]: Do you think we can have all clocks at same speed ? The ST Cube proposes this by default.

File: boards/nucleo-l476/include/periph_conf.h [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+#define SPI_NUMOF           (sizeof(spi_config) / sizeof(spi_config[0]))
+/** @} */
+
+/**
+ * @brief   ADC configuration
+ * @{
+ */
+#define ADC_NUMOF           (0)
+/** @} */
+
```
Comment (aabadie) [inline boards/nucleo-l476/include/periph_conf.h]: extra line

File: boards/nucleo-l476/include/periph_conf.h (hunk 11 above)
Comment (haukepetersen) [inline boards/nucleo-l476/include/periph_conf.h]: Actually, I think there are some constraints to the max clock values of the AHB buses (at least that was the case for other STM 32s)...

File: boards/nucleo-l476/include/periph_conf.h (hunk 12 above)
Comment (haukepetersen) [inline boards/nucleo-l476/include/periph_conf.h]: fixed

**Comment by aabadie** [comment]
there's a typo in the second (vor => for) commit message.

I tested on board, UART, LED, timer works. Good job, thanks :)

File: boards/nucleo-l476/include/periph_conf.h [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+ * @{
+ */
+static const pwm_conf_t pwm_config[] = {
+    {
+        .dev      = TIM2,
+        .rcc_mask = RCC_APB1ENR1_TIM2EN,
+        .chan     = { { .pin = GPIO_PIN(PORT_A, 15), .cc_chan = 0},
+                      { .pin = GPIO_PIN(PORT_B,  3), .cc_chan = 1},
+                      { .pin = GPIO_PIN(PORT_B, 10), .cc_chan = 2},
+                      { .pin = GPIO_PIN(PORT_B,  2), .cc_chan = 3} },
```
Comment (aabadie) [inline boards/nucleo-l476/include/periph_conf.h]: it should PORT_B, **11** on last channel

Comment (haukepetersen) [inline boards/nucleo-l476/include/periph_conf.h]: could you check the other nucleos (especially the nucleo-l1 and l0's), as I copied this from there somewhere...?! thanks

**Comment by haukepetersen** [comment]
Hmm, just found out, that the LSE is not working, more debugging to do...

File: boards/nucleo-l476/include/periph_conf.h (hunk 13 above)
Comment (aabadie) [inline boards/nucleo-l476/include/periph_conf.h]: check nucleo-l1 and l0's (L053 and 073) and they don't have this one...

**Comment by aabadie** [comment]
otherwise, I tested the UARTs and PWM (except the PB2) and all works.

One other thing: can you add this board to ARM_BOARDS variable in the unittest Makefile. I guess it will have enough memory.

**Comment by aabadie** [comment]
@haukepetersen, another thing, by adding this board, we complete the actual list of nucleo-64 boards supported by RIOT.

**Comment by haukepetersen** [comment]
Done, all good now with enabled LSE.

**Comment by aabadie** [comment]
Go !

**Pull request merged**